# Natural Language Processing

## 기본과제 3: Subword-level Language Model

> Reference 코드는 Solution 과 함께 공개됩니다.


### Introduction


* 본 과제의 목적은 서브워드 토큰화 (Subword Tokenization)의 필요성을 직접 느끼고 서브워드 토큰화 알고리즘 중 하나인 Byte Pair Encoding을 구현해봅니다.
* 서브워드 토큰화 기반 language model을 구현하면서 이전 과제의 Word-level language model과 비교해보는 시간을 갖겠습니다. 추가적으로 RNN을 LSTM으로 변경했을 때의 성능 차이에 대해 살펴보겠습니다.
* Subword-level language model을 구현하고, 주어진 데이터를 가공하여 모델을 학습한 후 학습된 언어 모델을 이용해 문장을 생성합니다.
* **ANSWER HERE** 이라고 작성된 부분을 채워 완성하시면 됩니다. 다른 부분의 코드를 변경하면 오류가 발생할 수 있습니다.

> 과제 완성 후 ipynb 파일을 제출해 주세요.<br>

### 0. 데이터 업로드


1. Boostcourse [기본 과제] Subword-level Language Model 에서 `wikitext-2.zip` 파일을 다운받습니다.
2. 본 Colab 환경에 `train.txt`, `dev.txt`, `test.txt` 파일을 업로드합니다.
3. `! ls` command 를 실행했을 때, `sample_data  test.txt  train.txt  valid.txt` 가 나오면 성공적으로 데이터 준비가 완료된 것 입니다.

In [2]:
! ls

'(10강-실습) HuggingFace_s Transformers 2.ipynb'
'(1강-실습-1) Naive Bayes classifier 구현.ipynb'
'(1강-실습-2) Corpus Cleaning.ipynb'
'(2강-실습-1) Word2Vec 구현 및 Embedding 시각화.ipynb'
'(2강-실습-2) 다국어 임베딩.ipynb'
'(3강-실습) Basic RNN 실습.ipynb'
'(4강-실습) LSTM, GRU 실습.ipynb'
'(5강-실습) Seq2Seq 구현.ipynb'
'(6강-실습) Seq2Seq with Attention 구현.ipynb'
'(7강-실습) Multi head Attention 구현.ipynb'
'(8강-실습) Masked Multi-head Attention 구현.ipynb'
'(9강-실습) HuggingFace_s Transformers 1.ipynb'
'(기본-1) Data Preprocessing_Tokenization (문제).ipynb'
'(기본-2) RNN-based Language Model (문제).ipynb'
'(기본-3) Subword-level Language Model (문제).ipynb'
'(기본-4) Preprocessing for NMT Model (문제).ipynb'
'(심화-1) BERT Fine-tuning with Transformers (문제).ipynb'
'(심화-2) NMT training with Fairseq (문제).ipynb'
 corpus.txt
 generate.txt
 generate_30epoch.txt
 generate_5epoch.txt
 model.pt
 model_30epoch.pt
 model_5epoch.pt
 test.txt
 train.txt
 valid.txt


In [3]:
path_train = './train.txt'
with open(path_train, 'r', encoding="utf8") as f:
    corpus_train = f.readlines()    

# train dataset 크기 확인
print(len(corpus_train))

# 처음 10 문장을 print 해 봅시다.
for sent_num, sent in enumerate(corpus_train[:10]):
    print(sent_num)
    print(sent)

36718
0
 

1
 = Valkyria Chronicles III = 

2
 

3
 Senjō no Valkyria 3 : <unk> Chronicles ( Japanese : 戦場のヴァルキュリア3 , lit . Valkyria of the Battlefield 3 ) , commonly referred to as Valkyria Chronicles III outside Japan , is a tactical role @-@ playing video game developed by Sega and Media.Vision for the PlayStation Portable . Released in January 2011 in Japan , it is the third game in the Valkyria series . <unk> the same fusion of tactical and real @-@ time gameplay as its predecessors , the story runs parallel to the first game and follows the " Nameless " , a penal military unit serving the nation of Gallia during the Second Europan War who perform secret black operations and are pitted against the Imperial unit " <unk> Raven " . 

4
 The game began development in 2010 , carrying over a large portion of the work done on Valkyria Chronicles II . While it retained the standard features of the series , it also underwent multiple adjustments , such as making the game more <unk> for ser

### 1. 서브워드 토큰화의 필요성

💡 서브워드(Subword)는 무엇인가요?

서브워드는 하나의 단어를 여러개의 단위로 분리했을 때 하나의 단위를 나타냅니다. `subword`를 서브워드 단위로 나타낸 하나의 예시는 다음과 같습니다.

 * `sub` + `word`

`sub`라는 접두사와 `word`라고 하는 어근으로 나누어 `subword`라고 하는 단어를 2개의 서브 워드로 나타냈습니다.

이외에도 다양한 형태의 서브워드로 나타낼 수 있습니다. (e.g., `su` + `bword`, `s` + `ubword`, `subwor` + `d`)


💡 그럼 서브워드 토큰화(Subword tokenization)는 무엇인가요?

서브워드 토큰화는 말 그대로 서브워드 단위로 토큰화를 한다는 뜻입니다.
기본 과제 1에서 나온 단어단위 토큰화를 적용한 뒤, 서브워드 토큰화를 수행한 예시를 보겠습니다.

서브워드 토큰화를 적용했을 때는 다음과 같이 토큰화할 수 있습니다.

* Example 1
> "I have a meal" -> ['I', 'hav', 'e', 'a', 'me', 'al']
>
> "나는 밥을 먹는다" -> ['나', '는', '밥', '을', '먹는', '다']

단어단위가 아니라 그보다 더 잘게 쪼갠 서브워드 단위로 문장을 토큰화합니다.

위에서 말씀드린 것과 같이 여러가지 경우의 수가 가능합니다.

* Example 2
> "I have a meal" -> ['I', 'ha', 've', 'a', 'mea', 'l']
>
> "나는 밥을 먹는다" -> ['나', '는', '밥', '을', '먹', '는다']

그렇지만 기본적으로 공백을 넘어선 서브를 구성하진 않습니다.
예를 들어 다음과 같이 토큰화를 수행하진 않습니다.

* Example 3
> "I have a meal" -> ['Iha', 've', 'am', 'ea', 'l']
>
> "나는 밥을 먹는다" -> ['나는밥', '을먹', '는다']

(참고4: [Huggingface: subword-tokenization](https://huggingface.co/transformers/tokenizer_summary.html#subword-tokenization))

💡 Subword tokenization은 왜 필요한가요?

첫 번째 이유는 이 세상에 단어가 너무 많기 때문입니다.
이전 과제에서 사용했던 코드를 불러와 그 필요성을 생각해 봅시다.

In [4]:
import os
import torch

class Dictionary(object):
    def __init__(self):
        self.token2id = {}
        self.id2token = []

    def add_word(self, word):
        if word not in self.token2id:
            self.id2token.append(word)
            self.token2id[word] = len(self.id2token) - 1
        return self.token2id[word]

    def __len__(self):
        return len(self.id2token)

class Corpus(object):
    def __init__(self, path):
        self.dictionary = Dictionary()
        self.train = self.tokenize(os.path.join(path, 'train.txt'))
        self.valid = self.tokenize(os.path.join(path, 'valid.txt'))
        self.test = self.tokenize(os.path.join(path, 'test.txt'))

    def tokenize(self, path):
        """Tokenizes a text file."""
        assert os.path.exists(path)
        # Add words to the dictionary
        with open(path, 'r', encoding="utf-8") as f:
            for line in f:
                words = line.split() + ['<eos>']
                for word in words:
                    self.dictionary.add_word(word)

        # Tokenize file content
        with open(path, 'r', encoding="utf-8") as f:
            idss = []
            for line in f:
                words = line.split() + ['<eos>']
                ids = []
                for word in words:
                    ids.append(self.dictionary.token2id[word])
                idss.append(torch.tensor(ids).type(torch.int64))
            ids = torch.cat(idss)

        return ids

In [5]:
from typing import Union, Tuple
import torch
import torch.nn as nn
import torch.nn.functional as F

class RNNModel(nn.Module):
    def __init__(self, 
        rnn_type: str,
        vocab_size: int,
        embedding_size: int=200,
        hidden_size: int=200,
        num_hidden_layers: int=2,
        dropout: float=0.5
    ):
        super().__init__()
        self.rnn_type = rnn_type
        self.vocab_size = vocab_size
        self.hidden_size = hidden_size
        self.num_hidden_layer = num_hidden_layers
        assert rnn_type in {'LSTM', 'GRU', 'RNN_TANH', 'RNN_RELU'}

        self.embedding = nn.Embedding(vocab_size, embedding_size)

        self.dropout = nn.Dropout(dropout)

        if rnn_type.startswith('RNN'):
            nonlinearity = rnn_type.split('_')[-1].lower()
            self.rnn = nn.RNN(
                embedding_size, 
                hidden_size, 
                num_hidden_layers,
                batch_first=True, 
                nonlinearity=nonlinearity,
                dropout=dropout
            )
        else:
            self.rnn = getattr(nn, rnn_type)(
                embedding_size,
                hidden_size,
                num_hidden_layers,
                batch_first=True,
                dropout=dropout
            )

        self.projection = nn.Linear(hidden_size, vocab_size)

    def forward(
        self, 
        input: torch.Tensor,
        prev_hidden: Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]
    ):
        ### YOUR CODE HERE (과제 2의 코드를 복사해서 넣으세요.)
        ### ANSWER HERE ###
            
        embeddings = self.embedding(input) # (batch_size, sequence_length, embedding_size)
        embeddings = self.dropout(embeddings)
        
        output, next_hidden = self.rnn(embeddings, prev_hidden) # (batch_size, sequence_length, hidden_size)
        output = self.dropout(output)
        
        projections = self.projection(output) # (batch_size, sequence_length, vocab_size)
        log_prob = F.log_softmax(projections, dim=-1) # (batch_size, sequence_length, vocab_size)

        ### END YOUR CODE
        
        assert list(log_prob.shape) == list(input.shape) + [self.vocab_size]
        assert prev_hidden.shape == next_hidden.shape if self.rnn_type != 'LSTM' \
          else prev_hidden[0].shape == next_hidden[0].shape == next_hidden[1].shape
        
        return log_prob, next_hidden
    
    def init_hidden(self, batch_size: int):
        weight = self.projection.weight
        
        if self.rnn_type == 'LSTM':
            return (weight.new_zeros(self.num_hidden_layer, batch_size, self.hidden_size),
                    weight.new_zeros(self.num_hidden_layer, batch_size, self.hidden_size))
        else:
            return weight.new_zeros(self.num_hidden_layer, batch_size, self.hidden_size)
    
    @property
    def device(self):   # 현재 모델의 device를 반환하는 프로퍼티
        return self.projection.weight.device

말뭉치의 문장들을 단어단위 토큰화를 해보고 단어들의 개수를 세어보겠습니다

In [6]:
corpus = Corpus('./')
vocab_size = len(corpus.dictionary)
print(vocab_size)

33278


이전 과제에 사용된 임베딩의 크기는 200이므로 단어 임베딩에 사용된 매개변수의 수는 33278 x 200 (6,655,600개)입니다.
그렇다면, RNN 모델에 사용되는 weight의 parameter 개수는 몇개인지 간단한 함수를 이용해 확인해보겠습니다

In [7]:
model = RNNModel('RNN_TANH', vocab_size)

In [8]:
def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)

print(f"Word embedding parameter 개수: {count_parameters(model.embedding)}")
print(f"RNN parameter 개수: {count_parameters(model.rnn)}")

Word embedding parameter 개수: 6655600
RNN parameter 개수: 160800


💡 RNN 층의 매개변수 개수와, 임베딩 매개변수 개수를 비교해보면 임베딩 매개변수의 개수가 RNN층의 매개변수 수보다 압도적으로 많습니다.

단어단위 임베딩을 사용하는 경우 학습에 사용되는 말뭉치의 크기가 커질수록 등장하는 단어가 더더욱 많아져 임베딩의 매개변수는 더 커지게 되고 전체 매개변수 대비 단어 임베딩이 차지하는 비중은 매우 높아집니다.

✨ 이런 매개변수 비중의 비대칭성을 해결하기 위해 처음에는 문자단위 토큰화(character-level tokenization) 방법이 주목을 받았습니다. 
말 그대로 하나의 글자를 기준으로 토큰화을 하는건데요.
이전 예시를 문자단위 토큰화를 하면 다음과 같습니다.

"I have a meal" -> ['I', 'h', 'a', 'v', 'e', 'a', 'm', 'e', 'a', 'l']
"나는 밥을 먹는다" -> ['나', '는', '밥', '을', '먹', '는', '다']

그러나, 문자단위 토큰화 역시 지나치게 긴 Sequence 길이, 성능 저하 등의 문제를 겪으며 서브워드 토큰화가 각광을 받게 되었습니다.




💡서브워드 토큰화가 가지는 두번째 장점은 Out-of-Vocabulary (OoV) 문제가 없다는 점입니다.

학습 데이터에서 등장하지 않은 단어는 모두 Unknown 토큰 [UNK]로 처리됩니다. 이는 테스트 과정 중에 처음 보는 단어를 모두 [UNK]로 모델의 입력을 넣게 되면서 전체적으로 모델의 성능이 저하될 수 있습니다.

그러나 서브워드 단위로 자르게 된다면 최악의 경우에도 문자단위로 토큰화가 진행됩니다. 이는 서브워드 토큰화는 현재 가지고 있는 Vocab으로 해당 단어가 토큰화할 수 없다면 그 단어를 서브워드 단위로 쪼개 평가하기 때문입니다.

따라서 서브워드 토큰화기는 가장 작은 문자 단위로 서브워드 토큰화가 가능하기 때문에 OoV 문제가 발생하지 않습니다. 

### 2. Byte Pair Encoding (BPE)
> 이 섹터에서는 파이썬 표준 라이브러리 (Python Standard Library)만 사용하세요.

대표적인 서브워드 토큰화 방법인 Byte pair encoding을 구현해봅시다. BPE의 정확한 알고리즘은 [논문](https://arxiv.org/pdf/1508.07909.pdf)의 3페이지 algorithm 1에 제시되어 있습니다. 각 문항과 주석의 지시사항을 확인하고 BPE를 구현해보세요.


### 2-A) BPE Vocab 만들기

BPE의 Vocab을 만드는 것은 간단합니다. 단순히 가장 많이 등장하는 연속한 짝을 찾아 추가하는 것 입니다.
다음과 같은 말뭉치가 있다고 가정해 봅시다.

```
low lower lowest newest
```

우선은 공백을 제외한 모든 문자를 Vocab에 추가하고 각 단어의 끝에 WORD_END "`_`" 붙여 단어를 구분지어 봅시다.

```
Vocab: d e i l n o r s t w _
[ l o w _ ], [ l o w e r _ ], [ l o w e s t _ ], [ w i d e s t _ ]  
```

이때 가장 많이 등장한 연속한 두 토큰을 찾아 Vocab에 추가하고 두 토큰을 붙입니다. 이 경우에는 `l o`가 세번 등장하여 가장 많았으니 `lo`로 붙여 Vocab에 추가합니다.

```
Vocab: d e i l n o r s t w _ lo
[ lo w _ ], [ lo w e r _ ], [ lo w e s t _ ], [ w i d e s t _ ] 
```

다음은 `lo w`가 세번 등장하므로 `low`를 추가합니다.

```
Vocab: d e i l n o r s t w _ lo low
[ low _ ], [ low e r _ ], [ low e s t _ ], [ w i d e s t _ ] 
```

다음은 `e s`가 두번 등장하므로 `es`를 추가합니다.

```
Vocab: d e i l n o r s t w _ lo low es
[ low _ ], [ low e r _ ], [ low es t _ ], [ w i d es t _ ] 
```

다음은 `es t`가 두번 등장하므로 `est`를 추가합니다.

```
Vocab: d e i l n o r s t w _ lo low es est
[ low _ ], [ low e r _ ], [ low est _ ], [ w i d est _ ] 
```

다음은 `est _`가 두번 등장하므로 `est_`를 추가합니다.

```
Vocab: d e i l n o r s t w _ lo low es est est_
[ low _ ], [ low e r _ ], [ low est_ ], [ w i d est_ ] 
```

`est_`는 est로 단어가 끝난다는 것을 알려주는 서브워드가 됩니다. 일반적으로 est가 나오면 단어가 끝나니 합리적입니다.

이러한 과정을 통해서 모든 단어가 추가되거나 원하는 Vocab 크기에 도달할 때까지 서브워드를 통합하여 추가하는 과정을 반복하면 됩니다. 알고리즘을 참고하여 `build_bpe`를 작성해 봅시다.

In [9]:
from typing import List
from collections import defaultdict, Counter
import re

# 단어 끝을 나타내는 문자
WORD_END = '_'

def build_bpe(
    corpus: List[str],
    max_vocab_size: int
) -> List[int]:
    """ BPE Vocab 만들기
    Byte Pair Encoding을 통한 Vocab 생성을 구현하세요.
    단어의 끝은 '_'를 사용해 주세요.
    이때 id2token을 서브워드가 긴 길이 순으로 정렬해 주세요.
    
    Note: 만약 모든 단어에 대해 BPE 알고리즘을 돌리게 되면 매우 비효율적입니다.
          왜냐하면 대부분의 단어는 중복되기 때문에 중복되는 단어에 대해서는 한번만 연산할 수 있다면 매우 효율적이기 때문입니다.
          따라서 collections 라이브러리의 Counter를 활용해 각 단어의 빈도를 구하고,
          각 단어에 빈도를 가중치로 활용하여 BPE를 돌리면 시간을 획기적으로 줄일 수 있습니다.
          물론 이는 Optional한 요소입니다.

    Arguments:
    corpus -- Vocab을 만들기 위한 단어 리스트
    max_vocab_size -- 최대 vocab 크기

    Return:
    id2token -- 서브워드 Vocab. 문자열 리스트 형태로 id로 token을 찾는 매핑으로도 활용 가능
    """
    ### YOUR CODE HERE
    ### ANSWER HERE ###
    corpus_sep = [word.replace("", " ")[1:] + WORD_END for word in corpus]
    corpus_counter = Counter(corpus_sep)
    vocab = list(set(' '.join(corpus_sep)))
    vocab.remove(' ')

    for _ in range(max_vocab_size - len(vocab)):
        pairs = defaultdict(int)
        for word, freq in corpus_counter.items():
            symbols = word.split()
            for i in range(len(symbols) - 1):
                pairs[symbols[i], symbols[i + 1]] += freq
        if not pairs:
            break
        
        best = max(pairs, key=pairs.get)
        vocab.append(''.join(best))
        
        new_vocab = {}
        bigram = re.escape(' '.join(best))
        p = re.compile(r'(?<!\S)' + bigram + r'(?!\S)')
        for word in corpus_counter:
            w_out = p.sub(''.join(best), word)
            new_vocab[w_out] = corpus_counter[word]
        corpus_counter = new_vocab
    
    id2token = sorted(vocab, key=len, reverse=True)
    ### END YOUR CODE
    
    return id2token

In [10]:
import re, collections

def get_stats(vocab):
    pairs = collections.defaultdict(int)
    for word, freq in vocab.items():
        symbols = word.split()
        for i in range(len(symbols)-1):
            pairs[symbols[i],symbols[i+1]] += freq 
    return pairs

def merge_vocab(pair, v_in):
    v_out = {}
    bigram = re.escape(' '.join(pair))
    p = re.compile(r'(?<!\S)' + bigram + r'(?!\S)') 
    for word in v_in:
        w_out = p.sub(''.join(pair), word)
        v_out[w_out] = v_in[word]
    return v_out

vocab = {'l o w </w>' : 5, 'l o w e r </w>' : 2, 
         'n e w e s t </w>': 6, 'w i d e s t </w>': 3}

num_merges = 15
for i in range(num_merges):
    pairs = get_stats(vocab)
    best = max(pairs, key=pairs.get) 
    vocab = merge_vocab(best, vocab) 
    print(best)
    print(vocab)

('e', 's')
{'l o w </w>': 5, 'l o w e r </w>': 2, 'n e w es t </w>': 6, 'w i d es t </w>': 3}
('es', 't')
{'l o w </w>': 5, 'l o w e r </w>': 2, 'n e w est </w>': 6, 'w i d est </w>': 3}
('est', '</w>')
{'l o w </w>': 5, 'l o w e r </w>': 2, 'n e w est</w>': 6, 'w i d est</w>': 3}
('l', 'o')
{'lo w </w>': 5, 'lo w e r </w>': 2, 'n e w est</w>': 6, 'w i d est</w>': 3}
('lo', 'w')
{'low </w>': 5, 'low e r </w>': 2, 'n e w est</w>': 6, 'w i d est</w>': 3}
('n', 'e')
{'low </w>': 5, 'low e r </w>': 2, 'ne w est</w>': 6, 'w i d est</w>': 3}
('ne', 'w')
{'low </w>': 5, 'low e r </w>': 2, 'new est</w>': 6, 'w i d est</w>': 3}
('new', 'est</w>')
{'low </w>': 5, 'low e r </w>': 2, 'newest</w>': 6, 'w i d est</w>': 3}
('low', '</w>')
{'low</w>': 5, 'low e r </w>': 2, 'newest</w>': 6, 'w i d est</w>': 3}
('w', 'i')
{'low</w>': 5, 'low e r </w>': 2, 'newest</w>': 6, 'wi d est</w>': 3}
('wi', 'd')
{'low</w>': 5, 'low e r </w>': 2, 'newest</w>': 6, 'wid est</w>': 3}
('wid', 'est</w>')
{'low</w>': 5,

**2-A 문제에 대한 테스트 코드**

In [11]:
print ("======Building BPE Vocab Test Case======")

# 첫번째 테스트
corpus = ['abcde']
vocab = build_bpe(corpus, max_vocab_size=15)
assert sorted(vocab, key=len, reverse=True) == vocab, \
       "id2token을 서브워드 길이가 긴 순으로 정렬해 주세요."
print("첫번째 테스트 통과!")

# 두번째 테스트
corpus = ['low'] * 5 + ['lower'] * 2 + ['newest'] * 6 + ['widest'] * 3
vocab = set(build_bpe(corpus, max_vocab_size=19))
assert vocab > {'est_', 'low', 'newest_', \
              'i', 'e', 'n', 't', 'd', 's', 'o', 'l', 'r', 'w', WORD_END} and \
       "low_" not in vocab and "wi" not in vocab and "id" not in vocab, \
       "BPE 결과가 기대한 결과와 다릅니다."
print("두번째 테스트 통과!")

# 세번째 테스트
corpus = ['aaaaaaaaaaaa', 'abababab']
vocab = set(build_bpe(corpus, max_vocab_size=8))
assert vocab == {'aaaaaaaa', 'aaaa', 'abab', 'aa', 'ab', 'a', 'b', WORD_END}, \
       "BPE 결과가 기대한 결과와 다릅니다."
print("세번째 테스트 통과!")

# 네번째 테스트
corpus = ['abc', 'bcd']
vocab = build_bpe(corpus, max_vocab_size=10000)
assert len(vocab) == 10, \
       "BPE 결과가 기대한 결과와 다릅니다."
print("네번째 테스트 통과!")

print("모든 테스트 통과!")

======Building BPE Vocab Test Case======
첫번째 테스트 통과!
두번째 테스트 통과!
세번째 테스트 통과!
네번째 테스트 통과!
모든 테스트 통과!


### 2-B) BPE 인코딩
만들어진 Vocab으로 텍스트 인코딩하는 방법은 몇 가지가 있습니다. 가장 일반적이고 정석인 방법은 앞에서부터 토큰화하되 가장 긴 것부터 욕심쟁이 기법(Greedy Search)으로 먼저 매칭하는 방법입니다.

```
Vocab: bcde ab cd bc de a b c d e _
abcde ==> ab cd e _
```

이 방법은 최적의 인코딩을 보장하진 않지만 긴 단어를 빠르게 인코딩하는 것이 가능합니다.

두번째 방법은 가장 길게 매칭되는 것을 전체 텍스트에 대해 먼저 토큰화하는 방법입니다.

```
Vocab: bcde ab cd bc de a b c d e _
abcde ==> a bcde _
```

두번째 방법은 첫번째 방법보다 느리지만 텍스트를 좀 더 짧게 인코딩하는 것이 가능합니다.

이 과제에서는 두번째 방법을 이용하여 BPE 인코딩을 구현해봅시다.

In [12]:
def encode(
    sentence: str,
    id2token: List[str]
) -> List[int]:
    """ BPE 인코더
    문장을 받아 BPE 토큰화를 통하여 고유 id의 수열로 바꿉니다.
    문장은 공백으로 단어단위 토큰화되어있다고 가정하며, Vocab은 sentence의 모든 문자를 포함한다고 가정합니다.
    찾을 수 있는 가장 긴 토큰부터 바꿉니다.
    
    Note: WORD_END를 빼먹지 마세요.

    Arguments:
    sentence -- 인코드하고자 하는 문장
    id2token -- build_bpe를 통해 만들어진 Vocab
    
    Return:
    token_ids -- 인코드된 토큰 id 수열
    """

    ### YOUR CODE HERE
    ### ANSWER HERE ###
    WORD_END = '_'
    sentence = sentence.replace(' ', WORD_END + ' ') + WORD_END
    words = list(sentence.split())
    
    token_ids = []
    for word in words:
        token_ids_tmp = []
        while word.split():
            for i, t in enumerate(id2token):
                if t in word:
                    token_ids_tmp.insert(word[:word.find(t)].count(' '), i)
                    word = word.replace(t, ' ', 1)
                    break
        token_ids += token_ids_tmp

    ### END YOUR CODE

    return token_ids

**2-B 문제에 대한 테스트 코드**

In [14]:
print ("======Encoding Test Case======")

# 첫번째 테스트
vocab = ['bcc', 'bb', 'bc', 'a', 'b', 'c', WORD_END]
assert encode('abbccc', vocab) == [3, 4, 0, 5, 6], \
       "BPE 인코딩 결과가 기대한 결과와 다릅니다."
print("첫번째 테스트 통과!")

print("모든 테스트 통과!")

# Second test
vocab = ['aaaa', 'aa', 'a', WORD_END]
assert len(encode('aaaaaaaa aaaaaaa', vocab)) == 7, \
       "BPE 인코딩 결과가 기대한 결과와 다릅니다."
print("두번째 테스트 통과!")

======Encoding Test Case======
첫번째 테스트 통과!
모든 테스트 통과!
두번째 테스트 통과!


### 2-C) BPE 디코딩

BPE로 인코딩된 것을 디코딩하는 것은 간단합니다.
그저 해당 id를 해당하는 서브워드로 만든 뒤 합치면됩니다.
WORD_END는 공백으로 처리하면 쉽습니다.

```
[ 196 62 20 6 ] ==> [ I_ li ke_ it_ ] ==> "I_like_it_" ==> "I like it " ==> "I like it"  
```

In [28]:
def decode(
    token_ids: List[int],
    id2token: List[str]
) -> str:
    """ BPE 디코더
    BPE로 토큰화된 id 수열을 받아 문장으로 바꿉니다.
    단어단위 토큰화에서의 문장 복원은 단순히 공백을 사이에 넣는 디코딩을 사용합니다.
    문장 끝의 공백은 잘라냅니다.
    
    Arguments:
    token_ids -- 디코드하고자하는 토큰 id 수열
    id2token -- build_bpe를 통해 만들어진 Vocab

    Return:
    sentence  -- 디코드된 문장
    """

    ### YOUR CODE HERE
    ### ANSWER HERE ###
    WORD_END = '_'
    sentence = ''
    token2id = {i: t for i, t in enumerate(id2token)}
    for ti in token_ids:
        sentence += token2id[ti]
    sentence = sentence.replace(WORD_END, ' ')
    sentence = sentence[:-1]

    ### END YOUR CODE
    return sentence

**2-C 문제에 대한 테스트 코드**

In [30]:
print ("======Decoding Test Case======")
# First test
vocab = ['bcc', 'bb', 'bc', 'a', 'b', 'c', WORD_END]
assert decode([3, 4, 0, 5, 6], vocab) == 'abbccc', \
       "BPE 디코딩 결과가 기대한 결과와 다릅니다."
print("첫번째 테스트 통과!")

# Second test
vocab = ['aaaa', 'aa', 'a', WORD_END]
assert decode([0, 0, 3, 0, 1, 2, 3], vocab) == 'aaaaaaaa aaaaaaa', \
       "BPE 디코딩 결과가 기대한 결과와 다릅니다."
print("두번째 테스트 통과!")

print('모든 테스트 통과!') 

======Decoding Test Case======
첫번째 테스트 통과!
두번째 테스트 통과!
모든 테스트 통과!


### 3. Transformers 라이브러리를 활용한 서브워드 토큰화

✨ 위에서 작성한 BPE 구현체를 통해 서브워드 토큰화의 원리를 알 수 있지만, 위의 구현체를 실제로 사용하기에는 난점이 존재합니다.
왜냐하면 BPE Vocab을 만드는 과정은 매우 오랜 시간이 걸리기 때문입니다.
다양한 토큰화기(tokenizer)를 직접 구현하고 학습하는 것은 매우 비용이 크기 때문에, 라이브러리를 활용하여 토큰화기를 사용하는 방법을 알아봅시다.

[Transformer](https://huggingface.co/docs/transformers/index) 라이브러리는 다양한 Transformer 구현체를 총망라한 라이브러리입니다.
Transfomer 외에도 다양한 토큰화기를 지원하는데, 이미 학습된 서브워드 토큰화기 역시 쉽게 불러올 수 있습니다.

(참고5: [Huggingface: subword tokenization](https://huggingface.co/transformers/tokenizer_summary.html#subword-tokenization))

In [31]:
# ! pip install transformers

In [34]:
from transformers import BertTokenizerFast

# BERT 모델에서 사용하는 토큰화를 가져옵니다.
# https://huggingface.co/docs/transformers/model_doc/bert#transformers.BertTokenizerFast
tokenizer = BertTokenizerFast.from_pretrained(
    "bert-base-cased",
    unk_token='<unk>',
    eos_token='<eos>'
)

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


**Question**

Transformers에서 제공되는 BertTokenizerFast는 모든 조합을 만들 수 있는 서브워드 기반 토큰화기임에도 불구하고 위와 같이 Unknown 토큰을 받을 수 있다. 서브워드 토큰화기에서 Unknown 토큰이 발생할 수 있는 상황은 무엇이 있을까?

**Answer**

1. 특수 문자 등 단어 자체가 subword 목록에 등록되지 않은 경우.

In [35]:
# 서브워드 토큰화 예시
print(tokenizer.tokenize('Boostcamp AI Tech'))
token_ids = tokenizer("Boostcamp AI Tech", add_special_tokens=False).input_ids
print(token_ids)
print(tokenizer.decode(token_ids))

print(tokenizer.tokenize('qwerklhfa asdfkwej'))
token_ids = tokenizer("qwerklhfa asdfkwej", add_special_tokens=False).input_ids
print(token_ids)
print(tokenizer.decode(token_ids))

['Bo', '##ost', '##cam', '##p', 'AI', 'Tech']
[9326, 15540, 24282, 1643, 19016, 7882]


2023-04-05 10:41:13.874729: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


Boostcamp AI Tech
['q', '##wer', '##k', '##l', '##h', '##fa', 'as', '##d', '##f', '##k', '##we', '##j']
[186, 12097, 1377, 1233, 1324, 8057, 1112, 1181, 2087, 1377, 7921, 3361]
qwerklhfa asdfkwej


In [39]:
words = 'qwerklhfa asdfkwej'
ids = tokenizer.encode(words)
print(ids)
print(tokenizer.decode(ids))

[101, 186, 12097, 1377, 1233, 1324, 8057, 1112, 1181, 2087, 1377, 7921, 3361, 102]
[CLS] qwerklhfa asdfkwej [SEP]


이 토큰화기는 `##`을 통하여 현 단어가 이전 단어와 연결되어 있는지를 알려주고 있습니다. 이 토큰화기를 기반으로 다시 모델을 선언하고 parameter의 개수를 살펴보겠습니다.

In [34]:
vocab_size = len(tokenizer)
subword_model = RNNModel('RNN_TANH', vocab_size)

In [35]:
print(f"임베딩 매개변수 개수: {count_parameters(subword_model.embedding)}")
print(f"RNN층 매개변수 개수: {count_parameters(subword_model.rnn)}")

임베딩 매개변수 개수: 5799600
RNN층 매개변수 개수: 160800


이전에 비해 임베딩 매개변수 개수는 확연히 줄어들었습니다.

6,655,600개 -> 5,799,600개

그에 비하여 이 토큰화기는 이전 토큰화기와 달리 학습 데이터에 없었던 영어 단어가 새로 나오더라도 토큰화가 가능합니다. 

그러면 이제 서브워드 토큰화 기반의 언어 모델 성능을 살펴봅시다.

### 4. 서브워드 기반 Language Model 학습
앞서 확인해본 `transformers` 라이브러리 기반 토큰화기를 활용하여 서브워드 기반 Lanuage Model을 학습시켜 봅시다. 기본 과제 2를 참고하여 구현해봅시다. 



In [4]:
### YOUR CODE HERE
### 기본 과제 2를 참고해서 Language model 학습 코드를 작성해 보세요.
### ANSWER HERE ###

In [5]:
import os
import math
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.nn.utils import clip_grad_norm_
from typing import Union, Tuple
from tqdm.notebook import tqdm
from transformers import BertTokenizerFast

In [6]:
class RNNModel(nn.Module):
    def __init__(self, 
        rnn_type: str,
        vocab_size: int,
        embedding_size: int=200,
        hidden_size: int=200,
        num_hidden_layers: int=2,
        dropout: float=0.5
    ):
        super().__init__()
        self.rnn_type = rnn_type
        self.vocab_size = vocab_size
        self.hidden_size = hidden_size
        self.num_hidden_layer = num_hidden_layers
        assert rnn_type in {'LSTM', 'GRU', 'RNN_TANH', 'RNN_RELU'}

        self.embedding = nn.Embedding(vocab_size, embedding_size)

        self.dropout = nn.Dropout(dropout)

        if rnn_type.startswith('RNN'):
            nonlinearity = rnn_type.split('_')[-1].lower()
            self.rnn = nn.RNN(
                embedding_size, 
                hidden_size, 
                num_hidden_layers,
                batch_first=True, 
                nonlinearity=nonlinearity,
                dropout=dropout
            )
        else:
            self.rnn = getattr(nn, rnn_type)(
                embedding_size,
                hidden_size,
                num_hidden_layers,
                batch_first=True,
                dropout=dropout
            )

        self.projection = nn.Linear(hidden_size, vocab_size)

    def forward(self, input: torch.Tensor, prev_hidden: Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]):
            
        embeddings = self.embedding(input) # (batch_size, sequence_length, embedding_size)
        embeddings = self.dropout(embeddings)
        
        output, next_hidden = self.rnn(embeddings, prev_hidden) # (batch_size, sequence_length, hidden_size)
        output = self.dropout(output)
        
        projections = self.projection(output) # (batch_size, sequence_length, vocab_size)
        log_prob = F.log_softmax(projections, dim=-1) # (batch_size, sequence_length, vocab_size)
        
        assert list(log_prob.shape) == list(input.shape) + [self.vocab_size]
        assert prev_hidden.shape == next_hidden.shape if self.rnn_type != 'LSTM' \
          else prev_hidden[0].shape == next_hidden[0].shape == next_hidden[1].shape
        
        return log_prob, next_hidden
    
    def init_hidden(self, batch_size: int):
        weight = self.projection.weight
        
        if self.rnn_type == 'LSTM':
            return (weight.new_zeros(self.num_hidden_layer, batch_size, self.hidden_size),
                    weight.new_zeros(self.num_hidden_layer, batch_size, self.hidden_size))
        else:
            return weight.new_zeros(self.num_hidden_layer, batch_size, self.hidden_size)
    
    @property
    def device(self):
        return self.projection.weight.device

In [7]:
def bptt_batchify(data: torch.Tensor, batch_size: int, sequence_length: int):
    num_sample = (len(data) // (batch_size * sequence_length))
    data = data[:num_sample * batch_size * sequence_length]
    batches = data.view((batch_size, num_sample, sequence_length)).transpose(0, 1)

    return batches

In [8]:
def train(model: RNNModel, data: torch.Tensor, lr: float):
    model.train()
    batch_size = data.shape[1]
    total_loss = 0.

    hidden = model.init_hidden(batch_size)
    progress_bar = tqdm(data, desc="Train")
    for bid, batch in enumerate(progress_bar, start=1):
        batch = batch.to(model.device)

        output, hidden = model(batch, hidden)
        if model.rnn_type == 'LSTM':
            hidden = tuple(tensor.detach() for tensor in hidden)
        else:
            hidden = hidden.detach()

        loss = F.nll_loss(output[:, :-1, :].transpose(1, 2), batch[:, 1:])
        
        model.zero_grad()
        loss.backward()

        clip_grad_norm_(model.parameters(), 0.25)
        for param in model.parameters():
            param.data.add_(param.grad, alpha=-lr)
        
        total_loss += loss.item()
        current_loss = total_loss / bid

        progress_bar.set_description(f"Train - loss {current_loss:5.2f} | ppl {math.exp(current_loss):8.2f} | lr {lr:02.2f}", refresh=False)

In [9]:
@torch.no_grad()
def evaluate(model: RNNModel, data: torch.Tensor):
    model.eval()
    
    total_loss = 0.
    hidden = model.init_hidden(data.shape[1])
    
    progress_bar = tqdm(data, desc="Evaluate")
    for bid, batch in enumerate(progress_bar, start=1):
        batch = batch.to(model.device)
        output, hidden = model(batch, hidden)
        if model.rnn_type == 'LSTM':
            hidden = tuple(tensor.detach() for tensor in hidden)
        else:
            hidden = hidden.detach()
        
        loss = F.nll_loss(output[:, :-1, :].transpose(1, 2), batch[:, 1:])
        total_loss += loss.item()
        current_loss = total_loss / bid
        
        progress_bar.set_description(f"Evaluate - loss {current_loss:5.2f} | ppl {math.exp(current_loss):8.2f}", refresh=False)
    
    loss = total_loss / bid

    return loss

In [10]:
tokenizer = BertTokenizerFast.from_pretrained("bert-base-cased", unk_token='<unk>', eos_token='<eos>')

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


In [11]:
batch_size = 16
sequence_length = 64

with open("./train.txt", 'r', encoding="utf8") as train_file:
    train_corpus = train_file.read()
train_data = torch.tensor(tokenizer(train_corpus, add_special_tokens=False).input_ids)
train_data = bptt_batchify(train_data, batch_size, sequence_length)
    
with open("./valid.txt", 'r', encoding="utf8") as valid_file:
    valid_corpus = valid_file.read()
valid_data = torch.tensor(tokenizer(valid_corpus, add_special_tokens=False).input_ids)
valid_data = bptt_batchify(valid_data, batch_size, sequence_length)
    
with open("./test.txt", 'r', encoding="utf8") as test_file:
    test_corpus = test_file.read()
test_data = torch.tensor(tokenizer(test_corpus, add_special_tokens=False).input_ids)
test_data = bptt_batchify(test_data, batch_size, sequence_length)

Token indices sequence length is longer than the specified maximum sequence length for this model (2284373 > 512). Running this sequence through the model will result in indexing errors


In [12]:
print(len(train_corpus))
print(len(valid_corpus))
print(len(test_corpus))

10780437
1120192
1255018


In [13]:
print(train_data.shape)
print(valid_data.shape)
print(test_data.shape)

torch.Size([2230, 16, 64])
torch.Size([222, 16, 64])
torch.Size([250, 16, 64])


In [14]:
train_data[0]

tensor([[  134, 12226,  3781,  ...,  1105,  3957,   119],
        [ 1563,   134,   134,  ..., 13539, 26626,  4362],
        [ 1343,  1150,  1209,  ...,   137,  4878,  4139],
        ...,
        [  157, 13144,  1731,  ...,   119, 13662,  1118],
        [  146,   112,  1396,  ...,  1111,  1760,  1776],
        [  107,   114, 23455,  ...,  1104, 16672,  1127]])

In [15]:
lr = 20.
num_epoch = 15 # 좀 더 그럴 듯한 결과를 원한다면 30 epoch 정도 돌리세요!
best_val_loss = None

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
vocab_size = len(tokenizer)
subword_model = RNNModel('RNN_TANH', vocab_size).to(device)

In [16]:
device

device(type='cuda')

In [ ]:
for epoch in range(1, num_epoch+1):
    train(subword_model, train_data, lr)
    val_loss = evaluate(subword_model, valid_data)
    print('-' * 89)
    print(f'| End of epoch {epoch:2d} | valid loss {val_loss:5.2f} | valid ppl {math.exp(val_loss):8.2f}')
    print('-' * 89)

    if not best_val_loss or val_loss < best_val_loss:
        torch.save(subword_model, 'subword_model_15epoch.pt')
        best_val_loss = val_loss
    else:
        lr /= 4.0

Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch  1 | valid loss  6.19 | valid ppl   490.03
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch  2 | valid loss  6.18 | valid ppl   481.87
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch  3 | valid loss  6.18 | valid ppl   483.54
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch  4 | valid loss  6.12 | valid ppl   455.91
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch  5 | valid loss  6.11 | valid ppl   450.54
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch  6 | valid loss  6.10 | valid ppl   443.97
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch  7 | valid loss  6.09 | valid ppl   442.20
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch  8 | valid loss  6.08 | valid ppl   438.91
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch  9 | valid loss  6.06 | valid ppl   429.63
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch 10 | valid loss  6.06 | valid ppl   429.67
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch 11 | valid loss  6.04 | valid ppl   420.19
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch 12 | valid loss  6.04 | valid ppl   418.05
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch 13 | valid loss  6.03 | valid ppl   417.52
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

Evaluate:   0%|          | 0/222 [00:00<?, ?it/s]

-----------------------------------------------------------------------------------------
| End of epoch 14 | valid loss  6.03 | valid ppl   416.14
-----------------------------------------------------------------------------------------


Train:   0%|          | 0/2230 [00:00<?, ?it/s]

In [ ]:
### END YOUR CODE

### 5. 학습한 언어 모델로 문장 생성

앞서 학습한 모델을 불러와서 문장을 생성해봅시다.
기본 과제 2와 똑같이 generate.txt를 저장하면 됩니다.


In [ ]:
### YOUR CODE HERE
### 기본 과제 2를 참고해서 Langauge model로 문장을 생성하는 코드를 작성해 보세요.
### ANSWER HERE ###

In [18]:
from tqdm.notebook import trange

subword_model = torch.load('subword_model_15epoch.pt', map_location=device)
subword_model.rnn.flatten_parameters()

test_loss = evaluate(subword_model, test_data)
print('=' * 89)
print(f'| End of training | test loss {test_loss:5.2f} | test ppl {math.exp(test_loss):8.2f}')
print('=' * 89)

num_words = 1000
temperature = 1.0

hidden = subword_model.init_hidden(1)
input = torch.randint(vocab_size, (1, 1), dtype=torch.long).to(device)
outputs = []

for i in trange(num_words, desc="Generation"):
    with torch.no_grad():
        log_prob, hidden = subword_model(input, hidden)
        
    weights = (log_prob.squeeze() / temperature).exp()
    token_id = torch.multinomial(weights, 1)
    outputs.append(token_id.item())
    input = token_id.unsqueeze(0)
    
outputs = tokenizer.decode(outputs)

with open("generation_subword_15epoch.txt", "w") as wf:
    wf.write(''.join(outputs).replace('<eos>', '\n'))

Evaluate:   0%|          | 0/250 [00:00<?, ?it/s]

| End of training | test loss  5.96 | test ppl   388.86


Generation:   0%|          | 0/1000 [00:00<?, ?it/s]

2023-04-05 11:41:52.622230: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [22]:
with open('./generation_subword_15epoch.txt', 'r', encoding="utf8") as rf:
    generation_test = rf.readlines()

In [24]:
generation_test

['in <unk> Sweet behavioruter release ). She and are March by the hero andtan of Kent, or a Storieswes of the other the listationh to Al. They death from November 2 and <unk> that Tasmanian is marked responsible / in Polish by his River. Although <unk>istic in 1958 military. They " In October from an short " functioning of the Pacific. Meanwhile a 1 is one " fix. It hasur by these and be called - with this Ta light. He thought the <unk> made structures S, while a case is recalled Grand at September whate grown which and once, which Depression it ingres made emerged Galaxy ; NY she, from the most is completely titled as a Parliament crash range of the role reason in the watershed District after a voyage @ - @ " secretary industries, the RAAF line to his failure update " I " and came " the last. He of the foot starts as a on <unk> Hot in the documentaries arrived extended fumbled " concluded and GA later\'s :, while. Throughout strategic defined the <unk> States Con. = = = Garner = = = =

In [ ]:
### END YOUR CODE

###**콘텐츠 라이선스**

<font color='red'><b>**WARNING**</b></font> : **본 교육 콘텐츠의 지식재산권은 재단법인 네이버커넥트에 귀속됩니다. 본 콘텐츠를 어떠한 경로로든 외부로 유출 및 수정하는 행위를 엄격히 금합니다.** 다만, 비영리적 교육 및 연구활동에 한정되어 사용할 수 있으나 재단의 허락을 받아야 합니다. 이를 위반하는 경우, 관련 법률에 따라 책임을 질 수 있습니다.
